### PR TITLE
fix: allign mono-repo structure - compilation works

### PR DESCRIPTION
**Pull request by ira-gordin-sap**

### CODE REVIEW

**Comment by tomer-epstein** [comment]
I'm expecting 'yarn' to run also the compilation.

**Comment by tomer-epstein** [comment]
After all changes approved, can you please squash all into 1 commit?

**Comment by ira-gordin-sap** [comment]
> I'm expecting 'yarn' to build all so developers can start to debug..... (instead of running 'yarn ci')

I think it is impossible. @bd82 could you please advise?

**Comment by ira-gordin-sap** [comment]
> After all changes approved, please squash all into 1 commit?

Currently this is one commit. I suppose I'll add more with code review fixes. And finally, I'll squash while merging.

**Comment by bd82** [comment]
> I think it is impossible. @bd82 could you please advise?

Probably possible with a `prepare` script or a `post-install` script.
But why is this needed? updating dependencies and running the build of the project are two separate
flows that may be executed separately...

File: .circleci/config.yml [hunk 1]
```diff
@@ -48,114 +52,80 @@ jobs: # a collection of steps
...
+            ghr -t ${GITHUB_TOKEN} -u ${CIRCLE_PROJECT_USERNAME} -r ${CIRCLE_PROJECT_REPONAME} -c ${CIRCLE_SHA1} -delete ${VERSION} ./artifacts/
 
 workflows:
   version: 2
-  test-deploy:
+  voter:
     jobs:
-      - build-and-test-node10
-      - build-and-test-node12
+      - build-node10
```
Comment (bd82) [inline .circleci/config.yml:108]: node10 is EOL in 3 weeks

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,26 +1,137 @@
...
+  yarn serve
+  ```
+- Open the broswer on `localhost:8080` to access the framework.
+
+#### Run the VSCode extension
+
+- Start VSCode on your local machine, and click on open workspace. Select this repo folder.
+- On the debug panel choose "Run Extension", and click on the "Run" button.
+
+#### Enable usage analytics reporting from VS Code
```
Comment (bd82) [inline CONTRIBUTING.md:85]: this analytics section does not belong in the contributing guide.
probably in the README of the VSCode ext

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,26 +1,137 @@
...
+
+The tool collects non-personally identifiable information about your usage of the tool to improve its services. If you do not want the tool to collect your usage data, you can set the "Enable Sap Web Analytics" setting to "false". Go to File > Preferences > Settings (macOS: Code > Preferences > Settings) > Extensions > Application Wizard, and deselect the "Enable Sap Web Analytics" checkbox.
+
+#### Advanced scenarios
```
Comment (bd82) [inline CONTRIBUTING.md:89]: does this section belong in contributing.md?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,26 +1,137 @@
...
+[jest]: https://jestjs.io/
+
+- To run the tests execute `yarn test` in a specific sub-package.
+- To run the tests with **coverage** run `yarn coverage` in a specific sub-package.
+
+### Code Coverage
+
+Code Coverage is enforced for all productive code in this mono repo.
+
+- Specific statements/functions may be [excluded][ignore_coverage] from the report.
```
Comment (bd82) [inline CONTRIBUTING.md:109]: are you using only jest? the documentation here links to istanbul ignore coverage docs, not jest...

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,26 +1,137 @@
...
+Performing a release requires push permissions to the repository.
+
+- Ensure you are on the default branch and synced with origin.
+- `yarn run release:version`
+- Follow the lerna CLI instructions.
+- Track the newly pushed **tag** (`/^v[0-9]+(\.[0-9]+)*/`) build in the build system
+  until successful completion.
+- Inspect the newly artifacts published on npmjs.com / Github Releases / other relevant release targets.
+
+## Contribute Code
```
Comment (bd82) [inline CONTRIBUTING.md:135]: I am not sure what is the purpose of this section

File: README.md [hunk 6]
```diff
@@ -11,59 +11,41 @@
...
+
+To report an issue please use the github issue tracker. Please try to make sure you have these in your issue:
+
+- No duplicate
+- Reproducible
+- Good summary
+- Well-documented
+- Minimal example
+
+## Issue handling process
```
Comment (bd82) [inline README.md:41]: I think there are templates you can add to .github directory to handle some of these docs related to issues.

File: lerna.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
+  "packages": ["packages/*"],
+  "useWorkspaces": true,
+  "npmClient": "yarn",
+  "command": {
+    "publish": {
+      "conventionalCommits": true
+    },
+    "version": {
```
Comment (bd82) [inline lerna.json:9]: use `allow-branch` option to safeguard the release flow
- https://github.com/lerna/lerna/tree/main/commands/version#--allow-branch-glob

File: nyc.config.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  reporter: ["text", "lcov"],
+  "check-coverage": true,
+  all: true,
+  // https://reflectoring.io/100-percent-test-coverage/
```
Comment (bd82) [inline nyc.config.js:5]: this comment does not seem to belong here
if you are not using enforcing 100% coverage

File: nyc.config.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+  branches: 86,
+  lines: 90,
+  functions: 88,
+  statements: 90,
+  // To enable **merged** coverage report all relevant file extensions must be listed.
+  extension: [".js", ".ts", ".vue"],
```
Comment (bd82) [inline nyc.config.js:11]: if you are using jest for vuejs testing you may not need the .vue
or maybe even the `.js` extensions here.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,99 @@
+{
+  "name": "root",
+  "private": true,
+  "workspaces": {
+    "packages": [
+      "packages/*"
+    ],
+    "nohoist": [
+      "**/@vue",
```
Comment (bd82) [inline package.json:9]: I do not get it,
why are you using `nohoist` for so many libraries?

File: package.json [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    "format:validate": "prettier --check \"**/*.@(js|ts|json|md|vue)\" --ignore-path=.gitignore",
+    "lint:fix": "eslint . --ext=js,ts,vue --fix --max-warnings=0 --ignore-path=.gitignore",
+    "lint:validate": "eslint . --ext=js,ts,vue --max-warnings=0 --ignore-path=.gitignore",
+    "coverage:clean": "rimraf ./coverage",
+    "coverage:merge": "lcov-result-merger \"./coverage/lcov*.info\" \"./coverage/lcov.info\"",
+    "legal:delete": "lerna exec \"shx rm -rf .reuse LICENSES\" || true",
+    "legal:copy": "lerna exec \"shx cp -r ../../.reuse .reuse && shx cp -r ../../LICENSES LICENSES\""
+  },
+  "dependencies": {
+    "lodash": "4.17.21"
```
Comment (bd82) [inline package.json:44]: do you have a dep to lodash here?
it is possible you no longer due if you have fewer scripts than the original ECMA mono repo template.

File: package.json [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    "nyc": "15.1.0",
+    "prettier": "2.2.1",
+    "typescript": "4.1.3",
+    "cz-conventional-changelog": "3.3.0",
+    "@commitlint/cli": "11.0.0",
+    "@commitlint/config-conventional": "11.0.0",
+    "rimraf": "3.0.2",
+    "coveralls": "2.11.16",
+    "shx": "0.3.3",
+    "patch-package": "^6.2.2",
```
Comment (bd82) [inline package.json:71]: i do not htink you use patch-package.

File: package.json [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    "lint:validate": "eslint . --ext=js,ts,vue --max-warnings=0 --ignore-path=.gitignore",
+    "coverage:clean": "rimraf ./coverage",
+    "coverage:merge": "lcov-result-merger \"./coverage/lcov*.info\" \"./coverage/lcov.info\"",
+    "legal:delete": "lerna exec \"shx rm -rf .reuse LICENSES\" || true",
+    "legal:copy": "lerna exec \"shx cp -r ../../.reuse .reuse && shx cp -r ../../LICENSES LICENSES\""
+  },
+  "dependencies": {
+    "lodash": "4.17.21"
+  },
+  "devDependencies": {
```
Comment (bd82) [inline package.json:46]: please validate that you actually need every single package.

File: packages/backend/.nycrc.json [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    "src/webSocketServer/*.ts",
+    "src/logUtils.ts",
+    "src/vscode-output.ts",
+    "src/youi-adapter.ts"
+  ],
+  "reporter": ["lcov", "text"],
+  "extension": [".ts"],
+  "all": true,
+  "check-coverage": true,
+  "branches": 85,
```
Comment (bd82) [inline packages/backend/.nycrc.json:18]: are these coverage thresholds duplicated from some root configuration?

File: packages/backend/README.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
+![GitHub license](https://img.shields.io/badge/license-Apache_2.0-blue.svg)
+
+# Application Wizard
+
+![](screenshot.png)
+
+With the Application Wizard extension, you can benefit from a rich user experience for yeoman generators. This extension allows developers to reuse existing yeoman generators and provide wizard-like experience with no development efforts.
+
+## Requirements
```
Comment (bd82) [inline packages/backend/README.md:9]: it seems you need **both** Theia and VSCode from this section

File: packages/backend/README.md [hunk 16]
```diff
@@ -0,0 +1,40 @@
+![GitHub license](https://img.shields.io/badge/license-Apache_2.0-blue.svg)
+
+# Application Wizard
+
+![](screenshot.png)
+
+With the Application Wizard extension, you can benefit from a rich user experience for yeoman generators. This extension allows developers to reuse existing yeoman generators and provide wizard-like experience with no development efforts.
```
Comment (bd82) [inline packages/backend/README.md:7]: put the short description text before the screenshot?

**Comment by bd82** [comment]
I think this change is a `refactor:` not a `fix:`

I reviewed what I could, but its very difficult to review like this with 170+ files changed.
The moving of files between directories should have been done in a separate PR to the refactoring to mono-repo structure.

File: packages/frontend/package.json [hunk 17]
```diff
@@ -1,31 +1,33 @@
...
     "@sap-devx/inquirer-gui": "0.1.26",
     "@sap-devx/inquirer-gui-file-browser-plugin": "0.0.5",
     "@sap-devx/inquirer-gui-folder-browser-plugin": "0.0.3",
+    "@sap-devx/inquirer-gui-label-plugin": "0.0.1",
     "@sap-devx/inquirer-gui-login-plugin": "0.0.4",
     "@sap-devx/inquirer-gui-tiles-plugin": "0.0.8",
-    "@sap-devx/inquirer-gui-label-plugin": "0.0.1",
     "@sap-devx/webview-rpc": "0.3.1",
-    "@sap-devx/yeoman-ui-types": "1.0.2",
+    "@sap-devx/yeoman-ui-types": "^1.1.60",
```
Comment (tomer-epstein) [inline packages/frontend/package.json:29]: pls, no ~.

**Review by tomer-epstein — CHANGES_REQUESTED** [review]
Run websocket fails

image

File: .circleci/config.yml (hunk 1 above)
Comment (ira-gordin-sap) [reply .circleci/config.yml:108]: @tomer-epstein, @rimasirich, may I remove node 10 build?

File: CONTRIBUTING.md (hunk 2 above)
Comment (ira-gordin-sap) [reply CONTRIBUTING.md:85]: This is already in README of the VSCode ext. Removing from here. Thanks!

File: CONTRIBUTING.md (hunk 4 above)
Comment (ira-gordin-sap) [reply CONTRIBUTING.md:109]: jest for frontend and nyc mocha for backend

File: package.json (hunk 10 above)
Comment (ira-gordin-sap) [reply package.json:9]: I had issues wih frontend build and functioning.

File: package.json (hunk 11 above)
Comment (ira-gordin-sap) [reply package.json:44]: Not sure I understand this comment.

**Comment by ira-gordin-sap** [comment]
> "typescript"

We have "typescript": "4.1.3" in the root package. Do you think the problem is the version? Maybe problem with changing the location. It can take time, so I prefer to do it in the separate PR.

File: packages/backend/README.md (hunk 16 above)
Comment (ira-gordin-sap) [reply packages/backend/README.md:5]: Thank you so much for all your comments and support while implementing this task. I fixed partially. Team decided to merge and do the rest of fixes in separate change.

**Comment by ira-gordin-sap** [comment]
We tested together and it works.

File: packages/backend/README.md (hunk 15 above)
Comment (ira-gordin-sap) [reply packages/backend/README.md:9]: Why? This is with or.
- [VSCode](https://code.visualstudio.com/) 1.39.2 or higher or [Theia](https://www.theia-ide.org/) 0.12 or higher.

**Review by tomer-epstein — APPROVED** [review]

File: README.md (hunk 6 above)
Comment (ira-gordin-sap) [reply README.md:41]: Great idea. We'll do it once I hope.

File: nyc.config.js (hunk 9 above)
Comment (ira-gordin-sap) [reply nyc.config.js:10]: It can stay here for the future maybe. And we run lcov-result-merger.

File: package.json (hunk 12 above)
Comment (ira-gordin-sap) [reply package.json:71]: You are right, but when I removed it yesterday vsix stopped working. I'll try to understand why.

**Pull request merged**